shared/params.go

```diff
@@ -269,28 +270,43 @@ func GetProductsForRequest(r *http.Request) (products []Product, err error) {
 	return products, nil
 }
 
-// ParseMaxCountParam parses the 'max-count' parameter as an integer, or returns 1 if no param
-// is present, or on error.
-func ParseMaxCountParam(r *http.Request) (count int, err error) {
-	return ParseMaxCountParamWithDefault(r, MaxCountDefaultValue)
-}
-
-// ParseMaxCountParamWithDefault parses the 'max-count' parameter as an integer, or returns the
-// default when no param is present, or on error.
-func ParseMaxCountParamWithDefault(r *http.Request, defaultValue int) (count int, err error) {
-	count = defaultValue
+// ParseMaxCountParam parses the 'max-count' parameter as an integer
+func ParseMaxCountParam(r *http.Request) (*int, error) {
 	if maxCountParam := r.URL.Query().Get("max-count"); maxCountParam != "" {
-		if count, err = strconv.Atoi(maxCountParam); err != nil {
-			return defaultValue, err
+		count, err := strconv.Atoi(maxCountParam)
+		if err != nil {
+			return nil, err
 		}
 		if count < MaxCountMinValue {
 			count = MaxCountMinValue
 		}
 		if count > MaxCountMaxValue {
 			count = MaxCountMaxValue
 		}
+		return &count, nil
+	}
+	return nil, nil
+}
+
+// ParseMaxCountParamWithDefault parses the 'max-count' parameter as an integer, or returns the
+// default when no param is present, or on error.
+func ParseMaxCountParamWithDefault(r *http.Request, defaultValue int) (count int, err error) {
+	if maxCountParam, err := ParseMaxCountParam(r); maxCountParam != nil || err != nil {
+		return *maxCountParam, err
```

Comment:
Won't this "segfault" if `err` is non-nil but `maxCountParam` is nil, as in the `return nil, err` case above? Is there a test for that case showing it not crashing? "http://wpt.fyi/?max-count=billions" I guess would do it?

Reply:
Fixeded